File: python/sum-of-multiples/sum_of_multiples.py
from itertools import count


def sum_of_multiples(limit: int, base_multiples: list[int]) -> int:
    base_multiples = [x for x in base_multiples if x != 0]
    if not base_multiples: return 0

    multiples: set[int] = set()

    for m in base_multiples:
        for n in count(1):
            product = m * n

            if product >= limit:
                break

            multiples.add(product)

    return sum(multiples)

File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import Data.List (group, sort)

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = sum uniqueMultiples
  where
    uniqueMultiples = removeDuplicates allFactors
    removeDuplicates = map head . group . sort
    allFactors = concatMap factorsUntilLimit factors
    factorsUntilLimit factor = factorsOf factor limit

factorsOf :: Integer -> Integer -> [Integer]
factorsOf 0 _ = []
factorsOf x limit = [0, x .. limit - 1]
